src/NzbDrone.Core/Download/History/DownloadHistoryService.cs

```diff
@@ -0,0 +1,164 @@
+                {
+                    return e;
+                }
+            }
+
+            return null;
+        }
+
+        public void Handle(EpisodeGrabbedEvent message)
+        {
+            _repository.Insert(new DownloadHistory
+            {
+                EventType = DownloadHistoryEventType.DownloadGrabbed,
+                SeriesId = message.Episode.Series.Id,
+                DownloadId = message.DownloadId,
+                SourceTitle = message.Episode.Release.Title,
+                Date = DateTime.UtcNow
+            });
+        }
+
+        public void Handle(EpisodeImportedEvent message)
+        {
+            if (!message.NewDownload)
+            {
+                return;
+            }
+
+            var downloadId = message.DownloadId;
+
+            if (downloadId.IsNullOrWhiteSpace())
+            {
+                downloadId = _historyService.FindDownloadId(message);
+            }
+
+            if (downloadId.IsNullOrWhiteSpace())
+            {
+                return;
+            }
+
+            _repository.Insert(new DownloadHistory
+            {
+                EventType = DownloadHistoryEventType.FileImported,
+                SeriesId = message.EpisodeInfo.Series.Id,
+                DownloadId = downloadId,
+                SourceTitle = message.EpisodeInfo.Path,
+                Date = DateTime.UtcNow
+            });
+        }
+
+        public void Handle(DownloadCompletedEvent message)
+        {
+            _repository.Insert(new DownloadHistory
+            {
+                EventType = DownloadHistoryEventType.DownloadImported,
+                SeriesId = message.TrackedDownload.RemoteEpisode.Series.Id,
+                DownloadId = message.TrackedDownload.DownloadItem.DownloadId,
+                SourceTitle = message.TrackedDownload.DownloadItem.OutputPath.ToString(),
+                Date = DateTime.UtcNow
+            });
+        }
+
+        public void Handle(DownloadFailedEvent message)
+        {
+            _repository.Insert(new DownloadHistory
+            {
+                EventType = DownloadHistoryEventType.DownloadFailed,
+                SeriesId = message.TrackedDownload.RemoteEpisode.Series.Id,
+                DownloadId = message.TrackedDownload.DownloadItem.DownloadId,
+                SourceTitle = message.SourceTitle,
+                Date = DateTime.UtcNow
+            });
+        }
+
+        public void Handle(DownloadIgnoredEvent message)
+        {
+            _repository.Insert(new DownloadHistory
+            {
+                EventType = DownloadHistoryEventType.DownloadIgnored,
+                SeriesId = message.SeriesId,
+                DownloadId = message.DownloadId,
+                SourceTitle = message.SourceTitle,
+                Date = DateTime.UtcNow
+            });
+        }
+
+        public void Handle(SeriesDeletedEvent message)
+        {
+            _repository.DeleteBySeriesId(message.Series.Id);
```

Comment:
As mentioned elsewhere, do we want to delete downloadHistory in this case? if so, probably better as Housekeeping job a la CleanupOrphanedHistoryItems.

Reply:
Unless we're not going to link it to the series, relink it to a series (making history mutable or series IDs consistent) I think it makes sense to clear it. Not deleting it here would be there is a point where we're inconsistent, the house keeping task for history was meant to ensure eventual consistency if handling the series delete failed, not as the only means. If we don't purge it here why clean it up at all?

Reply:
Hm, I thought we didn't clear History immediately on series deletions. But I just checked, we do, so I was wrong about that.

Reason for keeping downloadHistory is to be able to keep ignoring downloads for series that were already imported, but still seeding.
So that would be an argument to keep the DownloadHistory permanently.

Hm... If we tracked DownloadRemoved separately, then we could purge downloadhistory more selectively. That'd deal with the 'seeding' edge-case, while removing unused records.
I'm ok with doing that another time though.

Reply:
Yeah, it'd be nice to keep that all around, same as we talked about being able to map between episodes that change numbering so we can keep the linked episode files.